### PR TITLE
Support Client Assertion authentication on authentication client

### PR DESCRIPTION
**Pull request by ewanharris**
### 🔧 Changes

Introduces support for using Client Assertion  to the authentication client, this is for use when the Authentication Method is set to `Private Ket JWT`. This is supported using the `WithClientAssertion` option that can be passed instead of the existing `WithClientSecret`

```go
jwtPrivateKey := `-----BEGIN PRIVATE KEY-----
.......
-----END PRIVATE KEY-----`

api, err := New(
	context.Background(),
	domain,
	WithClientID(clientID),
	WithClientAssertion(jwtPrivateKey, "RS256"),
)

if err != nil {
	log.Fatal(err)
}

tokenSet, err := api.OAuth.LoginWithClientCredentials(context.Background(), oauth.LoginWithClientCredentialsRequest{}, oauth.IDTokenValidationOptions{})

if err != nil {
	log.Fatal(err)
}

log.Print(tokenSet)
```

### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### CODE REVIEW

File: authentication/http_recordings_test.go [hunk 1]
```diff
@@ -194,3 +216,14 @@ func redactDomain(i *cassette.Interaction, domain string) {
 	i.Response.Body = strings.ReplaceAll(i.Response.Body, domainParts[0], recordingsDomain)
 	i.Request.Body = strings.ReplaceAll(i.Request.Body, domainParts[0], recordingsDomain)
 }
+
+func verifyClientAssertion(t *testing.T, clientAssertion string) {
+	keyS := jwtPublicKey
+
+	key, err := jwk.ParseKey([]byte(keyS), jwk.WithPEM(true))
```
Comment (Widcket) [inline authentication/http_recordings_test.go:223]:
```suggestion
	keys := jwtPublicKey

	key, err := jwk.ParseKey([]byte(keys), jwk.WithPEM(true))
```

File: authentication/oauth.go [hunk 2]
```diff
@@ -193,21 +204,86 @@ func (o *OAuth) RevokeRefreshToken(ctx context.Context, body oauth.RevokeRefresh
...
+	case params.ClientAssertion != "":
+		body.Set("client_assertion", params.ClientAssertion)
+		body.Set("client_assertion_type", params.ClientAssertionType)
+		break
+	case clientSecret != "":
+		body.Set("client_secret", clientSecret)
+		break
+	}
+
+	if required && (body.Get("client_secret") == "" && body.Get("client_assertion") == "") {
```
Comment (Widcket) [inline authentication/oauth.go:242]: Should the `&&` in `(body.Get("client_secret") == "" && body.Get("client_assertion") == "")` be a `||`?

File: authentication/oauth.go [hunk 3]
```diff
@@ -193,21 +204,86 @@ func (o *OAuth) RevokeRefreshToken(ctx context.Context, body oauth.RevokeRefresh
...
+		return "", err
+	}
+
+	token, err := jwt.NewBuilder().
+		IssuedAt(time.Now()).
+		Subject(clientID).
+		JwtID(uuid.New().String()).
+		Issuer(clientID).
+		Audience([]string{domain}).
+		Expiration(time.Now().Add(2 * time.Minute)).
```
Comment (Widcket) [inline authentication/oauth.go:277]: Just out of curiosity, where does this 2 minute exp come from?

File: authentication/passwordless.go [hunk 4]
```diff
@@ -92,12 +105,27 @@ func (p *Passwordless) LoginWithSMS(ctx context.Context, params passwordless.Log
...
 }
 
-func (p *Passwordless) addClientAuthentication(params *oauth.ClientAuthentication) {
+func (p *Passwordless) addClientAuthentication(params *oauth.ClientAuthentication) error {
 	if params.ClientID == "" {
 		params.ClientID = p.authentication.clientID
 	}
 
-	if params.ClientSecret == "" && p.authentication.clientSecret != "" {
+	if p.authentication.clientAssertionSigningKey != "" {
```
Comment (Widcket) [inline authentication/passwordless.go:113]: Should we also check that `clientAssertionSigningAlg` is not empty?

File: authentication/http_recordings_test.go (hunk 1 above)
Comment (ewanharris) [reply authentication/http_recordings_test.go:223]: The reassigning here is a leftover before I extracted to a common key so I'll fix this to just remove `keyS`

File: authentication/oauth.go (hunk 2 above)
Comment (ewanharris) [reply authentication/oauth.go:242]: We want to error here if both values are empty as the they can be used independently of each other, so if `client_secret` is empty but `client_assertion` is not then we can continue

File: authentication/passwordless.go (hunk 4 above)
Comment (ewanharris) [reply authentication/passwordless.go]: Fixed in 4f1ea30418bb3789e920307da5e65652520252f1 (also changed in oauth.go)

File: authentication/http_recordings_test.go (hunk 1 above)
Comment (ewanharris) [reply authentication/http_recordings_test.go]: Fixed in 4f1ea30418bb3789e920307da5e65652520252f1

File: authentication/oauth.go (hunk 3 above)
Comment (ewanharris) [reply authentication/oauth.go:277]: I was aligning with the value set by the node-auth0 v4 beta here, I think there's not really a specific reason for picking 2 minutes, the only requirement is that `exp` must be a value no more than 5 minutes after the `iat`

Comment (ewanharris) [reply authentication/oauth.go:275]: I don't think we currently support configuring this on any other SDK and I haven't seen any request to do so yet, so I'd maybe defer to adding if we get a request

**Review by Widcket — APPROVED** [review]

File: authentication/authentication.go [hunk 5]
```diff
@@ -155,6 +158,7 @@ func New(ctx context.Context, domain string, options ...Option) (*Authentication
 		auth0ClientInfo:   client.DefaultAuth0ClientInfo,
 		idTokenSigningAlg: "RS256",
 		retryStrategy:     client.DefaultRetryOptions,
+		domain:            domain + "/",
```
Comment (sergiught) [inline authentication/authentication.go:161]: Should we use `url.JoinPath` here instead? If the domain already contains "/"?

File: authentication/passwordless.go [hunk 6]
```diff
@@ -70,7 +79,11 @@ func (p *Passwordless) SendSMS(ctx context.Context, params passwordless.SendSMSR
 //
 // See: https://auth0.com/docs/api/authentication?http#authenticate-user
 func (p *Passwordless) LoginWithSMS(ctx context.Context, params passwordless.LoginWithSMSRequest, validationOptions oauth.IDTokenValidationOptions, opts ...RequestOption) (t *oauth.TokenSet, err error) {
-	p.addClientAuthentication(&params.ClientAuthentication)
+	err = p.addClientAuthentication(&params.ClientAuthentication)
+
```
Comment (sergiught) [inline authentication/passwordless.go:83]:
```suggestion
```

File: authentication/passwordless_test.go [hunk 7]
```diff
@@ -135,3 +136,54 @@ func TestPasswordlessWithIDTokenVerification(t *testing.T) {
...
+
+		require.NoError(t, err)
+
+		r, err := api.Passwordless.SendSMS(context.Background(), passwordless.SendSMSRequest{
+			PhoneNumber: "+123456789",
+		})
+
+		assert.NoError(t, err)
+		assert.Equal(t, "+123456789", r.PhoneNumber)
+		assert.Equal(t, true, r.PhoneVerified)
```
Comment (sergiught) [inline authentication/passwordless_test.go:160]:
```suggestion
		assert.True(t, r.PhoneVerified)
```

File: authentication/passwordless_test.go [hunk 8]
```diff
@@ -135,3 +136,54 @@ func TestPasswordlessWithIDTokenVerification(t *testing.T) {
...
+			ClientAuthentication: oauth.ClientAuthentication{
+				ClientAssertion:     auth,
+				ClientAssertionType: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+			},
+			PhoneNumber: "+123456789",
+		})
+
+		assert.NoError(t, err)
+		assert.Equal(t, "+123456789", r.PhoneNumber)
+		assert.Equal(t, true, r.PhoneVerified)
```
Comment (sergiught) [inline authentication/passwordless_test.go:187]:
```suggestion
		assert.True(t, r.PhoneVerified)
```

**Review by sergiught — DISMISSED** [review]
Just very small suggestions, overall things look good 👍🏻

File: authentication/authentication.go (hunk 5 above)
Comment (ewanharris) [reply authentication/authentication.go:161]: Yeah that's a good point, on reflection it's probably a bit wasteful to have a separate domain here (just for the added `/`) when we already have url

**Review by sergiught — APPROVED** [review]

**Pull request merged**